### ISSUE TITLE
Training seems slow.... wondering if caused by "Converting sparse IndexedSlices to a dense Tensor" msg

### ISSUE THREAD

**Issue by davidsvaughn** [zhiguowang/BiMPM#5]
Training works, but it seems very slow... which is fine, as long as this is the expected behavior.  I'm just curious if it is unusually slow for me.... did you happen to get this warning?
```
/usr/local/lib/python2.7/dist-packages/tensorflow/python/ops/gradients_impl.py:93: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "
```
If not, then I'm wondering if my code is running slow due to this message... 
So far, I know it's coming from the `tf.gradients()` function call in line 220 of SentenceMatchModelGraph.py:
```
grads, _ = tf.clip_by_global_norm(tf.gradients(self.loss, tvars), clipper)
```
Unfortunately, this doesn't help much because I still don't know which part of the network is triggering it...  
If you tell me that your are not getting this message, then I will investigate further and try to find the root cause.  Thanks!

**Comment by ijinmao** [comment]
I implement a simplified keras version [here](https://github.com/ijinmao/BiMPM_keras). The training is also very slow. I found the slowest strategy is `Maxpooling-Matching` where "each forward (or backward) contextual embedding is compared with **every** forward (or backward) contextual embeddings of the other sentence". You could check it out by commenting each strategy in [`multi_perspective.py`](https://github.com/ijinmao/BiMPM_keras/blob/master/multi_perspective.py).

In my keras version, the warning is caused by `tf.gather()`. But I don't think this is the reason of slowness.

**Comment by fuhuamosi** [comment]
@davidsvaughn Hi, I notice that you said the training is very slow. Can you tell me the general training time of one epoch? Thank you.

**Comment by WangGangUCAS** [comment]
how to solve the problem?
